cache/client.go

```diff
@@ -0,0 +1,122 @@
+package cache
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/NebulousLabs/Sia/crypto"
+	"github.com/NebulousLabs/Sia/encoding"
+	"github.com/NebulousLabs/Sia/types"
+	"github.com/NebulousLabs/merkletree"
+)
+
+func verifyBlockHeader(header types.BlockHeader, minTimestamp types.Timestamp) error {
+	// Check that the timestamp is not too far in the past to be acceptable.
+	if minTimestamp > header.Timestamp {
+		return fmt.Errorf("Block header validation failed: EarlyTimestamp")
+	}
+
+	// Check if the block is in the extreme future. We make a distinction between
+	// future and extreme future because there is an assumption that by the time
+	// the extreme future arrives, this block will no longer be a part of the
+	// longest fork because it will have been ignored by all of the miners.
+	if header.Timestamp > types.CurrentTimestamp()+types.ExtremeFutureThreshold {
```

Comment:
Can you pass CurrentTimestamp as an argument? It would make testing easier.

Reply:
No, because the time passes between checks (`ExtremeFutureThreshold` and `FutureThreshold` checks). The main check (of target) will be added later on between those two checks, and the computation takes some time.

Reply:
In our case we already downloaded all headers from this node so the if it want to feed us with garbage it already succeeds. We need to check >150000 blocks very fast, so the time we spend on one block header should be really small (in Sia not only the header is checked so the time matters). I suggest to remove extra check and extra (possibly expensive) call to `CurrentTimestamp()`

Reply:
Just add a comment that check against types.ExtremeFutureThreshold is omitted because it does included in the further check and does not make sense as optimization in our case.

Reply:
Done.